### ISSUE TITLE
Invalid Union coercion order

### ISSUE THREAD

**Issue by mike0sv** [pydantic/pydantic#7022]
### Initial Checks

- [X] I have searched GitHub for a duplicate issue and I'm sure this is something new
- [X] I have searched Google & StackOverflow for a solution and couldn't find anything
- [X] I have read and followed [the docs](https://docs.pydantic.dev) and still think this is a bug
- [X] I am confident that the issue is with pydantic (not my code, or another library in the ecosystem like [FastAPI](https://fastapi.tiangolo.com) or [mypy](https://mypy.readthedocs.io/en/stable))

### Description

Because generic's `__getitem__` method is decorated with `lru_cache`, you can get different logic with the same code depending on order of execution. In the example below, second `List[Union[int, str]]` will actually return `List[Union[str, int]]` because it will hit the cache since `Union[str, int] == Union[int, str]`

I know about `smart_union`, but I actually rely on coercion order and I want `B(f=['1']).f[0] == 1`

I don't know if it applies to v2 since I didn't migrate yet - will it work there?

I get the feeling it will not be fixed in v1, but maybe there is some workaround? One I came up with is calling `List.__getitem__.__closure__[0].cell_contents.cache_clear()` before `B` declaration, but that looks ugly

### Example Code

```Python
from typing import List, Union

from pydantic import BaseModel


class A(BaseModel):
    f: List[Union[str, int]]


class B(BaseModel):
    f: List[Union[int, str]]


def main():
    print(type(B(f=[1]).f[0]))  # gives str instead of int


if __name__ == '__main__':
    main()
```

### Python, Pydantic & OS Version

```Text
pydantic version: 1.10.10
            pydantic compiled: False
                 install path: /Users/mike0sv/mambaforge/envs/evidently/lib/python3.9/site-packages/pydantic
               python version: 3.9.16 | packaged by conda-forge | (main, Feb  1 2023, 21:38:11)  [Clang 14.0.6 ]
                     platform: macOS-13.0.1-arm64-arm-64bit
     optional deps. installed: ['typing-extensions']
```

### Affected Components

- [ ] [Compatibility between releases](https://docs.pydantic.dev/changelog/)
- [X] [Data validation/parsing](https://docs.pydantic.dev/usage/models/#basic-model-usage)
- [ ] [Data serialization](https://docs.pydantic.dev/usage/exporting_models/) - `.dict()` and `.json()`
- [ ] [JSON Schema](https://docs.pydantic.dev/usage/schema/)
- [ ] [Dataclasses](https://docs.pydantic.dev/usage/dataclasses/)
- [ ] [Model Config](https://docs.pydantic.dev/usage/model_config/)
- [ ] [Field Types](https://docs.pydantic.dev/usage/types/) - adding or changing a particular data type
- [ ] [Function validation decorator](https://docs.pydantic.dev/usage/validation_decorator/)
- [X] [Generic Models](https://docs.pydantic.dev/usage/models/#generic-models)
- [ ] [Other Model behaviour](https://docs.pydantic.dev/usage/models/) - `construct()`, pickling, private attributes, ORM mode
- [ ] [Plugins](https://docs.pydantic.dev/) and integration with other tools - mypy, FastAPI, python-devtools, Hypothesis, VS Code, PyCharm, etc.

**Comment by PrettyWood** [comment]
```py
In [2] B.__annotations__
Out[2]: {'f': typing.List[typing.Union[str, int]]}
```
The issue comes from the order of the `Union` that is not preserved (by python not pydantic!) and your data is hence coerced

**Comment by mike0sv** [comment]
Yes, basically if you do `print(List[Union[int, str]], List[Union[str, int]])` you will get two `List[Union[int, str]]`.

So, v2 kinda has `smart_union` logic by default? What if I need the old logic, so that if I do `B(f=['1']).f[0]` I get and `int`?
